### DISCUSSION TITLE
How can I use GitHub Actions to automatically revert a PR if a later job fails? I want to automatically revert a pull request if a particular job in my GitHub Actions pipeline fails, even if previous jobs succeeded. What’s the best way to automate this rollback process?

### DISCUSSION THREAD

**Discussion by zeto-bit** [community/community#140805] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

How can I use GitHub Actions to automatically revert a PR if a later job fails?
I want to automatically revert a pull request if a particular job in my GitHub Actions pipeline fails, even if previous jobs succeeded. What’s the best way to automate this rollback process?

**Comment by AdarshBP** [comment, marked as answer]
To automatically revert a pull request (PR) in GitHub Actions if a later job fails, even if previous jobs succeeded, you can follow these steps:

1. Setup a Failing Job Check
You'll need to capture when a specific job fails. This can be done by checking the exit status or using GitHub's failure() function in your workflow.

2. Trigger a Revert
To revert the PR, you can use the GitHub API to:

Automatically create a new PR that reverts the changes made by the original PR.
Close or mark the original PR as reverted.
3. Configure a Personal Access Token
To perform the revert programmatically, GitHub Actions needs the correct permissions. You can create a Personal Access Token (PAT) with the necessary repository scopes and store it in GitHub Secrets.

4. GitHub Actions Workflow Example
Here’s an example of how to set up this logic:

name: PR with Auto-Revert

on:
  pull_request:
    branches:
      - main

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v3

      # Simulate a build step
      - name: Run some build steps
        run: |
          echo "Running build steps..."
          # Simulate failure in this step
          exit 1

  revert_on_failure:
    if: ${{ failure() }}
    runs-on: ubuntu-latest
    needs: build
    steps:
      - name: Checkout code
        uses: actions/checkout@v3

      # Revert the PR
      - name: Revert Pull Request
        run: |
          PR_NUMBER=$(echo ${{ github.event.number }})
          git fetch origin main
          git checkout -b revert-pr-$PR_NUMBER
          git revert -m 1 $(git log -1 --pretty=format:%H)
          git push origin revert-pr-$PR_NUMBER

      # Create a Revert PR using GitHub API
      - name: Create Revert PR
        uses: octokit/request-action@v2.x
        with:
          route: POST /repos/${{ github.repository }}/pulls
          token: ${{ secrets.GITHUB_TOKEN }}
          body: |
            {
              "title": "Revert PR #${{ github.event.number }}",
              "head": "revert-pr-${{ github.event.number }}",
              "base": "main",
              "body": "Automatically reverting due to failure in the build job."
            }
